lib/puppet/util/autoload.rb

```diff
@@ -163,7 +162,9 @@ def gem_directories
     end
 
     def search_directories(env)
-      [gem_directories, module_directories(env), libdirs(), $LOAD_PATH].flatten
+      # Ensure that module_directories have been added to $LOAD_PATH
+      module_directories(env)
```

Comment:
Another question I have - is the precedent order what we want? I.e. what happens in collisions?

You would know better than I how the search is handled - is code loaded from earlier paths first with the ability to reopen / redefine code of the same type loaded later?

Will this in any way impact module code that performs feature checks against core Puppet types? (ACL module for instance will monkey patch classes based on the Puppet version - it checks for defined methods IIRC)